src/features/ExtensionCommands.ts

```diff
@@ -382,24 +388,45 @@ export class ExtensionCommandsFeature implements IFeature {
         return promise;
     }
 
-    private saveFile(filePath: string): Thenable<EditorOperationResponse> {
+    private async saveFile(saveFileDetails: ISaveFileDetails): Promise<EditorOperationResponse> {
 
-        let promise: Thenable<EditorOperationResponse>;
-        if (this.findTextDocument(this.normalizeFilePath(filePath))) {
-            promise =
-                vscode.workspace.openTextDocument(filePath)
-                    .then((doc) => {
-                        if (doc.isDirty) {
-                            doc.save();
-                        }
-                    })
-                    .then((_) => EditorOperationResponse.Completed);
-        } else {
-            promise = Promise.resolve(EditorOperationResponse.Completed);
+        // If the file to save can't be found, just complete the request
+        if (!this.findTextDocument(this.normalizeFilePath(saveFileDetails.filePath))) {
+            return EditorOperationResponse.Completed;
         }
 
-        return promise;
-    }
+        // If no newFile is given, just save the current file
+        if (!saveFileDetails.newPath) {
+            const doc = await vscode.workspace.openTextDocument(saveFileDetails.filePath);
+            if (doc.isDirty) {
+                await doc.save();
+            }
+
+            return EditorOperationResponse.Completed;
+        }
+
+        // Otherwise we want to save as a new file
+
+        // First turn the path we were given into an absolute path
+        // Relative paths are interpreted as relative to the original file
+        const newFileAbsolutePath = path.isAbsolute(saveFileDetails.newPath) ?
+            saveFileDetails.newPath :
+            path.resolve(path.dirname(saveFileDetails.filePath), saveFileDetails.newPath);
+
+        // Retrieve the text out of the current document
+        const oldDocument = await vscode.workspace.openTextDocument(saveFileDetails.filePath);
+        const oldDocText = oldDocument.getText();
+
+        // Write it to the new document path
+        fs.writeFileSync(newFileAbsolutePath, oldDocText);
+
+        // Finally open the new document
+        const newFileUri = vscode.Uri.file(newFileAbsolutePath);
+        const newFile = await vscode.workspace.openTextDocument(newFileUri);
+        vscode.window.showTextDocument(newFile, { preview: false });
```

Comment:
Why did you choose to open it not as a preview? Just curious. Not really leaning either way personally.

Reply:
Ah, didn't know what it was (just followed the response on SO figuring it prevented some behaviour). I feel like maybe it actually should be a preview...

Reply:
I'm leaning on preview actually too 😄